[LLM] Add docs for serving Llama3 (#3449)
skypilot-org/skypilot: llm/llama-3/README.md

<!-- $REMOVE -->
# Scale Serving Llama-3 on Any Cloud or Kubernetes with SkyPilot

> <!-- $END_REMOVE -->

Reviewer: Consider having a nice "Llama3 + SkyPilot" banner too, or an AI-generated image of 3 llamas?

Author: Added one. Thanks!

<!-- $UNCOMMENT# Llama-3: Open LLM from Meta -->

> [Llama-3](https://github.com/meta-llama/llama3) is the latest top open-source models. It has been released with a license that authorizes commercial use. You can deploy a private Llama-3 chatbot with SkyPilot in your own cloud with just one simple command.

Reviewer: ```suggestion
[Llama-3](https://github.com/meta-llama/llama3) is the latest top open-source LLM from Meta. It has been released with a license that authorizes commercial use. You can deploy a private Llama-3 chatbot with SkyPilot in your own cloud with just one simple command.
```

* [Llama-3 release](https://github.com/meta-llama/llama3)
* [Llama-3 blog](https://ai.meta.com/blog/meta-llama-3/)

<p align="center">
<img src="https://imgur.com/zPpY2Bg.gif" alt="Gradio UI serving Llama-3" style="width: 80%;">
</p>

> ## Why use SkyPilot to deploy over commercial hosted solutions?

Reviewer: ```suggestion
## Why use SkyPilot vs. commercial hosted solutions?
```

* No lock-in: run on any supported cloud - AWS, Azure, GCP, Lambda Cloud, IBM, Samsung, OCI
* Everything stays in your cloud account (your VMs & buckets)
* No one else sees your chat history
* Pay absolute minimum — no managed solution markups
* Freely choose your own model size, GPU type, number of GPUs, etc, based on scale and budget.

…and you get all of this with 1 click — let SkyPilot automate the infra.

## Prerequisites

- Go to the [HuggingFace model page](https://huggingface.co/meta-llama/Meta-Llama-3-70B-Instruct) and request access to the model `meta-llama/Meta-Llama-3-70B-Instruct`.
- Check that you have installed SkyPilot ([docs](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)).
- Check that `sky check` shows clouds or Kubernetes are enabled.

## SkyPilot YAML

<details>
<summary>Click to see the full recipe YAML</summary>

> ```yaml

Reviewer: Seems like some differences from the actual YAML content.

Author: Good catch! Updated. Thanks!

I 04-18 16:34:13 optimizer.py:909]  AWS          p4de.24xlarge               96      1152      A100-80GB:8    us-east-1          40.97               
I 04-18 16:34:13 optimizer.py:909] ------------------------------------------------------------------------------------------------------------------
...
```

</details>

Wait until the model is ready (this can take 10+ minutes), as indicated by these lines:
```console
...
(task, pid=17433) Waiting for vllm api server to start...
...
(task, pid=17433) INFO:     Started server process [20621]
(task, pid=17433) INFO:     Waiting for application startup.
(task, pid=17433) INFO:     Application startup complete.
(task, pid=17433) INFO:     Uvicorn running on http://0.0.0.0:8081 (Press CTRL+C to quit)
...
(task, pid=17433) Running on local URL:  http://127.0.0.1:8811
(task, pid=17433) Running on public URL: https://xxxxxxxxxx.gradio.live
...
(task, pid=17433) INFO 03-28 04:32:50 metrics.py:218] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%
```

> :tada: **Congratulations!** :tada: You have now launched the Llama-3 Instruct LLM on your infra.

Reviewer: Nit: this didn't render on the DBRX page. Maybe check https://skypilot.readthedocs.io/en/latest/gallery/frameworks/ollama.html#serving-llama2-with-a-cpu-instance to see how that tip icon was rendered on HTML.

Author: Good catch! Updated.

You can play with the model via
- Standard OpenAPI-compatible endpoints (e.g., `/v1/chat/completions`)
- Gradio UI (automatically launched)

To curl `/v1/chat/completions`:
```console
IP=$(sky status --ip llama3)

# We need to manually specify the stop_token_ids to make sure the model finish

> # on <|eot_id|>.

Reviewer: Q: will this be fixed by vllm transparently?

Author: Currently, no. It has to be specified for now. vllm updated their code to load generation_config.json to update the stop ids, but meta has not updated their llama released config yet.

llama3          1   1        xx.yy.zz.121  18 mins ago  1x GCP([Spot]{'A100-80GB': 4})  READY   us-east4
llama3          2   1        xx.yy.zz.245  18 mins ago  1x GCP([Spot]{'A100-80GB': 4})  READY   us-east4
```
</details>

Get a single endpoint that load-balances across replicas:
```console
ENDPOINT=$(sky serve status --endpoint llama3)
```

> **Tip:** SkyServe fully manages the lifecycle of your replicas. For example, if a spot replica is preempted, the controller will automatically replace it. This significantly reduces the operational burden while saving costs.

To curl the endpoint:
```console
curl -L $ENDPOINT/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "meta-llama/Meta-Llama-3-70B-Instruct",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Who are you?"
      }
    ]
  }'
```

To shut down all resources:
```console
sky serve down llama3
```

See more details in [SkyServe docs](https://skypilot.readthedocs.io/en/latest/serving/sky-serve.html).

> ### **Optional**: Connect a GUI to your Llama-3 endpoint

Reviewer: Maybe put the GUI gif here as well?

Author: It might be a bit duplicated and may have too many moving items in the page which may be a bit distracting. How about I put a skypilot+llama3 banner on the top and move that demo back here?

Reviewer: SGTM.

It is also possible to access the Code Llama service with a GUI using [FastChat](https://github.com/lm-sys/FastChat).

1. Start the chat web UI:
```bash
sky launch -c llama3-gui ./gui.yaml --env ENDPOINT=$(sky serve status --endpoint llama3)
```

2. Then, we can access the GUI at the returned gradio link:
```
| INFO | stdout | Running on public URL: https://6141e84201ce0bb4ed.gradio.live
```

Note that you may get better results to use a higher temperature and top_p value.

> ## Finetunning Llama-3

Reviewer: ```suggestion
## Finetuning Llama-3
```

<!-- $REMOVE -->
# Scale Serving Llama-3 on Any Cloud or Kubernetes with SkyPilot
<!-- $END_REMOVE -->
<!-- $UNCOMMENT# Llama-3: Open LLM from Meta -->

<p align="center">

> <img src="https://imgur.com/fOCr394.png" alt="Llama-3 x SkyPilot" style="width: 60%;">

Reviewer: Consider making the whole "logo + skypilot" a bit smaller, leaving some white space within the bottom box. Logo needs to be smaller in proportion to the skypilot text too :)

Author: I made it smaller and reduce the size of skypilot logo. How does it look now?
![](https://imgur.com/BwY6PUQ.png)

Reviewer: Maybe even smaller? Some whitespace is good
